src/toolkits/object_detection/od_darknet_yolo_model.cpp

```diff
@@ -0,0 +1,399 @@
+/* Copyright © 2020 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#include <toolkits/object_detection/od_darknet_yolo_model.hpp>
+
+#include <toolkits/object_detection/od_yolo.hpp>
+
+namespace turi {
+namespace object_detection {
+
+namespace {
+
+using neural_net::compute_context;
+using neural_net::float_array_map;
+using neural_net::image_augmenter;
+using neural_net::model_spec;
+using neural_net::Publisher;
+using neural_net::shared_float_array;
+using neural_net::xavier_weight_initializer;
+
+using padding_type = model_spec::padding_type;
+
+// The spatial reduction depends on the input size of the pre-trained model
+// (relative to the grid size).
+// TODO: When we support alternative base models, we will have to generalize.
+constexpr int SPATIAL_REDUCTION = 32;
+
+constexpr float BASE_LEARNING_RATE = 0.001f;
+
+// Each bounding box is evaluated relative to a list of pre-defined sizes.
+const std::vector<std::pair<float, float>>& GetAnchorBoxes() {
+  static const std::vector<std::pair<float, float>>* const default_boxes =
+      new std::vector<std::pair<float, float>>({
+          {1.f, 2.f},
+          {1.f, 1.f},
+          {2.f, 1.f},
+          {2.f, 4.f},
+          {2.f, 2.f},
+          {4.f, 2.f},
+          {4.f, 8.f},
+          {4.f, 4.f},
+          {8.f, 4.f},
+          {8.f, 16.f},
+          {8.f, 8.f},
+          {16.f, 8.f},
+          {16.f, 32.f},
+          {16.f, 16.f},
+          {32.f, 16.f},
+      });
+  return *default_boxes;
+};
+
+// These are the fixed values that the Python implementation currently passes
+// into TCMPS.
+// TODO: These should be exposed in a way that facilitates experimentation.
+// TODO: A struct instead of a map would be nice, too.
+
+float_array_map GetBaseBackendConfig() {
+  float_array_map config;
+  config["learning_rate"] = shared_float_array::wrap(BASE_LEARNING_RATE);
+  config["gradient_clipping"] = shared_float_array::wrap(0.025f);
+  // TODO: Have MPS path use these parameters, instead
+  // of the values hardcoded in the MPS code.
+  config["od_rescore"] = shared_float_array::wrap(1.0f);
```

Comment:
Should we change config to a struct specific to OD?

Reply:
Yeah, that's something I'd like to do. Should probably be a separate PR though. Ultimately we should decide whether we actually will ever expose these as user-controllable parameters. If not, they can be implementation details of individual model backends